lang/LangPrimSource/PyrUnixPrim.cpp

```diff
@@ -147,35 +152,29 @@ static void string_popen_thread_func(struct sc_process *process)
 int prString_POpen(struct VMGlobals *g, int numArgsPushed);
 int prString_POpen(struct VMGlobals *g, int numArgsPushed)
 {
-	struct sc_process *process;
 	PyrSlot *a = g->sp - 1;
 	PyrSlot *b = g->sp;
-	int err;
 
 	if (!isKindOfSlot(a, class_string)) return errWrongType;
 
-	char *cmdline = (char*)malloc(slotRawObject(a)->size + 1);
-	err = slotStrVal(a, cmdline, slotRawObject(a)->size + 1);
-	if(err) {
-		free(cmdline);
-		return errFailed;
-	}
+	char *cmdline = new char[slotRawObject(a)->size + 1];
+	slotStrVal(a, cmdline, slotRawObject(a)->size + 1);
 
 #ifdef SC_IPHONE
 	SetInt(a, 0);
 	return errNone;
 #endif
 
-	process = (struct sc_process *)malloc(sizeof(struct sc_process));
+	sc_process *process = new sc_process;
```

Comment:
what about:
`std::unique_ptr<sc_process> process ( new sc_process )`

if `string_popen_thread_func` takes a smart pointer as argument, you can model the transfer of ownership of `sc_process` to the helper thread. no need to free the struct manually